include/ginkgo/core/reorder/rcm.hpp

```diff
@@ -0,0 +1,196 @@
+#include <ginkgo/core/reorder/reordering_base.hpp>
+
+
+namespace gko {
+/**
+ * @brief The Reorder namespace.
+ *
+ * @ingroup reorder
+ */
+namespace reorder {
+
+enum class starting_strategy { minimum_degree, pseudo_peripheral };
+
+
+/**
+ * Rcm is the reordering algorithm which uses METIS to compute a
+ * fill reducing re-ordering of the given sparse matrix. The METIS
+ * `Metis_NodeND` which is implements the multilevel nested dissection algorithm
+ * as in the METIS documentation.
+ *
+ * @note  This class is derives from polymorphic object but is not a LinOp as it
+ * does not make sense for this class to implement the apply methods. The
+ * objective of this class is to generate a reordering/permutation vector (in
+ * the form of the Permutation matrix), which can be used to apply to reorder a
+ * matrix as required.
+ *
+ * <Explanation starting strategy here>
+ *
+ * @tparam ValueType  Type of the values of all matrices used in this class
+ * @tparam IndexType  Type of the indices of all matrices used in this class
+ *
+ * @ingroup reorder
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Rcm : public EnablePolymorphicObject<Rcm<ValueType, IndexType>,
+                                           ReorderingBase> {
+    friend class EnablePolymorphicObject<Rcm, ReorderingBase>;
+
+public:
+    using SparsityMatrix = matrix::SparsityCsr<ValueType, IndexType>;
+    using PermutationMatrix = matrix::Permutation<IndexType>;
+
+    /**
+     * Gets the adjacency matrix representation of the system operator (input
+     * matrix) of the linear operator.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const SparsityMatrix> get_adjacency_matrix() const
+    {
+        return adjacency_matrix_;
+    }
+
+    /**
+     * Gets the permutation (permutation matrix, output of the algorithm) of the
+     * linear operator.
+     *
+     * @return the permutation (permutation matrix)
+     */
+    std::shared_ptr<const PermutationMatrix> get_permutation() const
+    {
+        return permutation_;
+    }
+
+    /**
+     * Gets the inverse permutation (permutation matrix, output of the
+     * algorithm) of the linear operator.
+     *
+     * @return the inverse permutation (permutation matrix)
+     */
+    std::shared_ptr<const PermutationMatrix> get_inverse_permutation() const
+    {
+        return inv_permutation_;
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * If this parameter is set then an inverse permutation matrix is also
+         * constructed along with the normal permutation matrix.
+         */
+        bool GKO_FACTORY_PARAMETER(construct_inverse_permutation, false);
+
+        /**
+         * This parameter controls the strategy used to determine a starting
+         * vertex.
+         */
+        starting_strategy GKO_FACTORY_PARAMETER(
+            starting_strategy, starting_strategy::pseudo_peripheral);
```

Comment:
I think you would need to change the name of the factory parameter or the enum as they are [clashing right now](https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/517071998#L543).